src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/gpt_model.py

```diff
@@ -0,0 +1,179 @@
+from torch import Tensor
+
+from megatron.bridge.models.qwen_vl.modelling_qwen3_vl.rope import (
+    Qwen3VLMoETextRotaryEmbedding,
+    Qwen3VLTextRotaryEmbedding,
+)
+from megatron.bridge.models.qwen_vl.modelling_qwen3_vl.transformer_block import Qwen3VLTransformerBlock
+from megatron.bridge.models.transformer_config import TransformerConfig
+
+
+class Qwen3VLGPTModel(GPTModel):
+    """Qwen3-VL GPT model with vision-language capabilities."""
+
+    def __init__(
+        self,
+        config: TransformerConfig,
+        transformer_layer_spec: ModuleSpec,
+        vocab_size: int,
+        max_sequence_length: int,
+        pre_process: bool = True,
+        post_process: bool = True,
+        fp16_lm_cross_entropy: bool = False,
+        parallel_output: bool = True,
+        share_embeddings_and_output_weights: bool = False,
+        position_embedding_type: Literal["learned_absolute", "rope", "mrope", "none"] = "learned_absolute",
+        rotary_percent: float = 1.0,
+        rotary_base: int = 10000,
+        rope_scaling: bool = False,
+        rope_scaling_factor: float = 8.0,
+        scatter_embedding_sequence_parallel: bool = True,
+        seq_len_interpolation_factor: Optional[float] = None,
+        mtp_block_spec: Optional[ModuleSpec] = None,
+        vp_stage: Optional[int] = None,
+    ) -> None:
+        super().__init__(
+            config=config,
+            transformer_layer_spec=transformer_layer_spec,
+            vocab_size=vocab_size,
+            max_sequence_length=max_sequence_length,
+            pre_process=pre_process,
+            post_process=post_process,
+            fp16_lm_cross_entropy=fp16_lm_cross_entropy,
+            parallel_output=parallel_output,
+            share_embeddings_and_output_weights=share_embeddings_and_output_weights,
+            position_embedding_type=position_embedding_type,
+            rotary_percent=rotary_percent,
+            rotary_base=rotary_base,
+            rope_scaling=rope_scaling,
+            rope_scaling_factor=rope_scaling_factor,
+            scatter_embedding_sequence_parallel=scatter_embedding_sequence_parallel,
+            seq_len_interpolation_factor=seq_len_interpolation_factor,
+            mtp_block_spec=mtp_block_spec,
+            vp_stage=vp_stage,
+        )
+
+        is_moe = (
+            hasattr(config, "num_moe_experts") and config.num_moe_experts is not None and config.num_moe_experts > 0
+        )
+
+        if is_moe:
+            self.rotary_pos_emb = Qwen3VLMoETextRotaryEmbedding(config.hf_text_config)
+        else:
+            self.rotary_pos_emb = Qwen3VLTextRotaryEmbedding(config.hf_text_config)
+
+        self.mrope_section = self.config.mrope_section
+        assert self.mrope_section is not None, (
+            "mrope require mrope_section setting, but we got None from TransformerConfig"
+        )
+
+        # rebuild the transformer block
+        self.decoder = Qwen3VLTransformerBlock(
```

Comment:
shouldn't need rebuild blocks. It's extra overhead, just update the layer spec to use Qwen3VLTransformerBlock